### PR TITLE
test(versions): Restrict Node versions allowed in JavaScript category

### PR DESCRIPTION
**Pull request by ExE-Boss**
This ensures that in the JavaScript category, only the Node release that introduced a new V8 version is allowed.

## Caveats:
- Node.js v8.5.0 introduced the `‑‑experimental‑modules` flag, but didn’t ship a new V8 version, so it has to be special cased.
- Node.js v13.0.0 completed I18n support and introduced experimental support for `WeakRef`s behind the `‑‑harmony‑weak‑refs` flag, but didn’t ship a new V8 version, so it also has to be special cased.

---

~~I’ve also added the Node versions from #3160.~~

### CODE REVIEW

**Comment by vinyldarkscratch** [comment]
I haven't really looked too deep into this PR, but right off the bat, I can see that this includes some changes that are beyond the scope of the title, such as the addition of Node versions, and code styling changes that have no relevance or alteration in functionality.  Can we make sure that we're cherry-picking changes into their own individual PRs?

**Comment by freaktechnik** [comment]
The inclusion of the node versions was actually requested in https://github.com/mdn/browser-compat-data/pull/3160#issuecomment-498265988 to be in this change.

**Comment by vinyldarkscratch** [comment]
Huh, that seems odd for them to request.  Personally, I would expect two separate PRs -- one for the versions, and one for the test -- to get a better-looking commit log.

Regardless, there are the formatting alterations that I feel should be omitted, both since they don't change the functionality of the program and all the other tests follow along with the existing formatting style.  Not that I'm against them in general, but it makes the git blame a little cleaner.  😛

**Comment by ExE-Boss** [comment]
> I haven't really looked too deep into this PR, but right off the bat, I can see that this includes some changes that are beyond the scope of the title, such as the addition of Node versions, and code styling changes that have no relevance or alteration in functionality.

All code changes are required by this, **eg.** the addition of `hasCategorySpecificData` is needed to display a better error message for `nodejs` versions in the `javascript` category.

Similarly, I don’t want to re‑do the (somewhat expensive) `getValidBrowserVersions(…)` call for every `SimpleSupportStatement`, so I cache the result at the first moment it’s required.

**Comment by freaktechnik** [comment]
Pretty sure the "code style changes" are a reference to the chalk coloring changes.

File: test/test-versions.js [hunk 1]
```diff
@@ -101,23 +151,28 @@ function testVersions(filename) {
...
+          category ? category : prop,
+        );
       }
     }
   }
   findSupport(data);
 
   if (errors.length) {
-    console.error(chalk`{red   Versions – }{red.bold ${errors.length}}{red  ${errors.length === 1 ? 'error' : 'errors'}:}`);
+    console.error(chalk`{red   Versions – {bold ${errors.length}} ${errors.length === 1 ? 'error' : 'errors'}:}`);
```
Comment (vinyldarkscratch) [inline test/test-versions.js]: I'm talking about things like this.  Changes like this one that do not adjust the functionality, especially based upon the title of the PR, should be in separate PRs.

**Comment by Elchi3** [comment]
This is really hard to review as it adds even more complexity to test-versions. I'd prefer if we first think about making test-versions less of a monolithic function and have tests for it. Otherwise, this will need an in-depth reviewing for both the new functionality and to make sure nothing is regressed.

**Comment by ddbeck** [comment]
Another thought on this: can we break out the new Node.js data itself? #3160 has been blocked on that for some time.

**Review by alshamiri5 — APPROVED** [review]

**Comment by saschanaz** [comment]
I don't think this is always valid, for example Node 14.8.0 didn't introduce a new V8 version but it did enable unflagged top level await. https://nodejs.org/en/blog/release/v14.8.0/

**Comment by ddbeck** [comment]
This has been open a long time and seems to have never really come up as a data quality issue. Given https://github.com/mdn/browser-compat-data/pull/4264#issuecomment-722525633, this probably can't be enforced without continuing to monitor for exceptions, too. For both of those reasons, I'm going to close it. Nevertheless, thank you for your effort here!
